### PR TITLE
Pass `return_type` to `AccumulatorFunctionImplementation ` for user defined aggregates

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3427

 # Rationale for this change
See https://github.com/apache/arrow-datafusion/issues/3427

# What changes are included in this PR?
- [x] Pass`return_type` in `AccumulatorFunctionImplementation`
- [x] Improve some comments

# Are there any user-facing changes?
Yes, if you use user defined aggregate functions -- one of the type signatures has changed

### CODE REVIEW

File: datafusion/core/src/physical_plan/udaf.rs [hunk 1]
```diff
@@ -99,7 +100,7 @@ impl AggregateExpr for AggregateFunctionExpr {
     }
 
     fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
-        (self.fun.accumulator)()
+        (self.fun.accumulator)(&self.data_type)
```
Comment (alamb) [inline datafusion/core/src/physical_plan/udaf.rs:103]: The is the key change -- pass the return type when instantiating the accumulator

File: datafusion/expr/src/function.rs [hunk 2]
```diff
@@ -39,15 +39,16 @@ use std::sync::Arc;
...
+/// Factory that returns the functions's return type given the input argument types
 pub type ReturnTypeFunction =
     Arc<dyn Fn(&[DataType]) -> Result<Arc<DataType>> + Send + Sync>;
 
-/// the implementation of an aggregate function
+/// Factory that returns an accumulator for the given aggregate, given
+/// its return datatype.
 pub type AccumulatorFunctionImplementation =
-    Arc<dyn Fn() -> Result<Box<dyn Accumulator>> + Send + Sync>;
+    Arc<dyn Fn(&DataType) -> Result<Box<dyn Accumulator>> + Send + Sync>;
```
Comment (alamb) [inline datafusion/expr/src/function.rs:49]: Her is the signature change

**Comment by alamb** [comment]
This will likely conflict with https://github.com/apache/arrow-datafusion/pull/3425 logically

**Review by andygrove — APPROVED** [review]
Looks reasonable to me. Thanks @alamb

**Pull request merged**